Add an API to add a plugin sidebar (new)

IreneStr opened the pull request:
## Description
This adds an API that plugins can use to add sidebars.

The API looks like this:



```js
wp.editPost.registerSidebar( 
	"my-plugin/my-sidebar", 
	{
		title: "Name",
		render: () => { return <div>My sidebar</div>;}
	} 
);
```
```js
wp.editPost.activateSidebar( "my-plugin/my-sidebar" );
```

<!-- Please describe your changes -->

## How Has This Been Tested?
There is a PR on Yoast SEO to test this with. Yoast/wordpress-seo#8531 shows how the API can be used.

## Screenshots (jpeg or gifs if applicable):
<img width="1474" alt="schermafbeelding 2018-02-14 om 14 29 35" src="https://user-images.githubusercontent.com/17744553/36206667-894f8ec8-1193-11e8-8a48-aba2ca050f8f.png">

## Types of changes
New API

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

## Remarks
- When triggering a plugin sidebar from outside Gutenberg (see the Yoast branch mentioned by @atimmer above), there is a timing issue. Sometimes the activation happens before the registration, which results in an error `No matching plugin sidebar found for plugin [plugin name]`. When refreshing the page with the sidebar opened, the same error will appear.
- Because of we're handling sidebars differently now, mobileMiddleware is no longer needed. Whether the mobile sidebar should be opened is now based on a `viewport` property in the state (it used to be based on `sidebar: 'mobile'` in the state). I've removed the mobile middleware in the `edit-post` folder. It would be great if someone can double check whether I didn't break anything related to that. 
- The sidebar is hidden when you change from desktop to mobile. However, it's not saved that it was opened before, which means it will not be reopened when switching back to desktop. 

For more comments, see https://github.com/WordPress/gutenberg/pull/4109

IreneStr committed:
Make it possible for plugins to render a sidebar

IreneStr committed:
Fix sidebar styling

IreneStr committed:
Remove unused mobile middleware

IreneStr committed:
Fix editor mode switching

IreneStr committed:
Fix editorMode switch test

IreneStr committed:
Fix documentation

IreneStr committed:
Merge branch 'master' into add/api-add-plugin-sidebar-2

atimmer committed:
Change `@returns` into `@return`

jasmussen commented:
I think I said this on the last PR, but I'll say it again: nice work — I think developers will love this API.

Keep in mind I can only speak for the UI and user experienceon this, so take any thoughts from me in that vein.

The principles are these:

- A plugin can register a sidebar
- This sidebar can be opened from the More menu, if the user clicks the item there, or a toolbar icon button, if the sidebar is pinned
- A toolbar icon is pinned, but can be unpinned. I previously suggested that a plugin can't itself pin the icon, but the user has to do this. However per the consensus in our recent slack chats, let's double back on that, and let a plugin add a toolbar icon if it wants to. But the user should still be able to _unpin_ that icon (see the star bookmark button in older mockups).
- A sidebar with an unpinned icon can be opened from the More menu item.
- A pinned icon is a single-color inline SVG icon

As also mentioned in a previous PR, not all of these principles have to be part of THIS PR. It's totally fine that we address these later on. But I'm explicitly stating these so we can have some agreement as to where this should end up.

If we agree this is the approach to pursue and this will be possible and desirable, then I think UI wise this PR is good to go after two quick visual fixes. The plugin sidebar should look a bit more like the standard sidebar:

![screen shot 2018-02-08 at 09 53 21](https://user-images.githubusercontent.com/1204802/35964413-f9c634c6-0cb7-11e8-966b-e1fd5ea9435c.png)

Right now the plugin sidebar has a bigger font size in the title (should be `font-size: 100%;`), the titlebar is 56px, should be 50px, and the close icon is right-aligned, should have a 10px right margin. In other words this is how it should look:

![screen shot 2018-02-08 at 09 55 20](https://user-images.githubusercontent.com/1204802/35964462-19c03b8c-0cb8-11e8-8770-11b6fdd0e306.png)

jasmussen commented:
Another quick note I forgot to mention — the behavior for the sidebar hiding and not restoring when you resize the screen has regressed from master. We should ideally restore it like we do in master. 

We also need to be sure that the mobile experience is good, the modal sidebar system has regressed also:

![screen_shot_2018-02-08_at_12 17 42](https://user-images.githubusercontent.com/1204802/35970582-27389468-0ccb-11e8-8ea1-60a77e89cf8b.png)

But modals on mobile need imminent polish regardless, as part of https://github.com/WordPress/gutenberg/issues/4082, so perhaps that doesn't need to be addressed in this PR. The thing is — due to how WordPress handles fixed and sticky toolbars at various breakpoints, and due to the need to handle mobile scroll bleed (don't scroll content underneath the modal when you're swiping, and also don't lose your place when you open a modal), it's difficult to get the modals to behave well here. I know @youknowriad has a component that takes some of the vinegar out.

jorgefilipecosta commented on edit-post/store/selectors.js:
```diff
@@ -16,8 +16,18 @@ export function getEditorMode( state ) {
+ * @return {string}        Active plugin sidebar plugin
+ */
+export function getActivePlugin( state ) {
+	return getPreference( state, 'activeSidebarPanel', {} ).plugins;
```
Minor: in the state tree we save as "plugins", but the selector uses "Plugin". Should we allow more than one plugin to be active? If yes the selector should be renamed to getActivePlugins if not the property should be renamed to plugin.

IreneStr replied:
@jorgefilipecosta That's a good question. I think because this function is sidebar specific (using 'activeSidebarPanel' in the getPreference function), there could only be one plugin active at any given time. I'll rename the property accordingly.

jorgefilipecosta commented on edit-post/components/plugins-panel/index.js:
```diff
@@ -0,0 +1,80 @@
+export default connect(
+	( state ) => {
+		return {
+			plugin: getActivePlugin( state ),
```
It looks like this will always return null (and we will see the 'No matching plugin sidebar found for plugin ""', because the store being queried here is not the store that contains the active ActiveEditorPanel preference.
We can confirm that by adding console.log( 'PluginsPanel' , state ); before the return,.
We can correct that by specifying the store key:
```
export default connect(
	( state ) => {
		return {
			plugin: getActivePlugin( state ),
		};
	}, {
		onClose: closeGeneralSidebar,
	}, undefined,
	{ storeKey: 'edit-post' }
)( withFocusReturn( PluginsPanel ) );
```

jorgefilipecosta reviewed (commented):
It looks like the changes, to the mobile sidebar, hiding/persisting behavior are good, nice job in that refactoring 👍

jorgefilipecosta commented on edit-post/store/defaults.js:
```diff
@@ -1,9 +1,10 @@
-		publish: false,
+	editorMode: 'visual',
+	viewportType: 'desktop', // 'desktop' | 'mobile'
+	activeGeneralSidebar: null, // null | 'editor' | 'plugins'
```
Looks like here we should set activeGeneralSidebar to 'editor', otherwise we are introducing a behavior change because when reloading the page no sidebar will be active.

gziolo reviewed (commented):
Thank you for your work on this PR. It looks like we are pretty close to have it merged.

I think we still need to clarify how the public API should look like. I left some question regarding `renderSidebar` and `getSidebarSettings` methods exposed. It seems like `registerSidebar` would be a better fit here. I'm wondering what's your take on this and why it has changed in this new PR.

I played with the code locally and I saw some issues:

![screen shot 2018-02-08 at 11 58 40](https://user-images.githubusercontent.com/699132/35972491-59ceb608-0cd2-11e8-8ead-08be452ce6f0.png)

I think @jorgefilipecosta found the reason of this issue. We need to fix it before merging this PR. The screenshot I shared reveals also the issue with the `renderSidebar` method - which registers the sidebar every time it's going to be executed. This needs to be optimized.

I have a feeling that it might be possible to reuse or extract more functionality out of the existing sidebar components. This might be done in the follow-up PRs, we just should identify what would be the best way to move forward.

Another thing to think about in the future (not for this PR) is if we could handle the global settings and block settings as sidebar plugins, too. This would allow us further simplify the way we handle state because we wouldn't have to worry about two competing sidebars which want to be displayed in the same area. 

Let's make sure there are no unexpected behaviors and come up with the plan how to continue work on this feature.

jorgefilipecosta commented:
We are persisting the active plugin panel, so if when we reload because of some reason the sidebar is not available (e.g: plugin was uninstalled) we immediately enter into an error state. As a future improvement when loading the preferences from localStorage we should check if the sidebar is available before showing the error. That is not a blocker we may improve this part after the merge :)

jorgefilipecosta commented:
Thank you a lot for the work in this PR, I left some suggestions about possible improvements/bugfixes but generally, it looks like we are getting close to a merge :)

gziolo commented on edit-post/components/plugins-panel/index.js:
```diff
@@ -0,0 +1,80 @@
+				/>
+			</div>
+			<div className="editor-plugins-panel__content">
+				{ render() }
```
On second thought I don't think we need to change that. It is perfectly fine as long as we make it clear that you need to return `WPElement` from the function, similar to what [createElement](https://github.com/WordPress/gutenberg/blob/master/element/index.js#L27) produces:

```js
const render = () => <MyComponent />;
```

IreneStr committed:
Add storeKey to make the plugins panel connect to the correct store

IreneStr committed:
Merge branch 'master' into add/api-add-plugin-sidebar-2

IreneStr committed:
Rename 'plugins' to 'plugin' in selector

IreneStr committed:
Set editor as default activeGeneralSidebar

IreneStr commented on edit-post/api/index.js:
```diff
@@ -0,0 +1,4 @@
+export {
+	renderSidebar,
+	getSidebarSettings,
```
The exposure of getSidebarSettings was a remnant of some debugging. I'll remove it from the API. I cannot think of a reason to leave it exposed. When someone thinks of a use case it can be exposed again.

IreneStr committed:
Unexpose getSidebarSettings

IreneStr commented on edit-post/api/sidebar.js:
```diff
@@ -0,0 +1,114 @@
+ */
+function activateSidebar( name ) {
+	store.dispatch( openGeneralSidebar( 'plugins' ) );
+	store.dispatch( setGeneralSidebarActivePanel( 'plugins', name ) );
```
This will be addressed in a next iteration

IreneStr commented on edit-post/api/sidebar.js:
```diff
@@ -0,0 +1,114 @@
+		return null;
+	}
+
+	settings = applyFilters( 'editor.registerSidebar', settings, name );
```
This will be addressed in a next iteration

IreneStr committed:
Remove duplicate scss and remove renderSidebar function

IreneStr commented on edit-post/components/plugins-panel/index.js:
```diff
@@ -0,0 +1,80 @@
+			tabIndex="-1">
+			<div className="editor-plugins-panel__header">
+				<h3>{ title }</h3>
+				<IconButton
```
The close button is the only similarity. The header in sidebar/header.js solely consists of three buttons. The header component may be generalized to fit both headers, but that's something for a next iteration.

IreneStr commented on edit-post/components/plugins-panel/index.js:
```diff
@@ -0,0 +1,80 @@
+		render,
+	} = getPluginSidebar( plugin );
+	return (
+		<div
```
Based on my comment https://github.com/WordPress/gutenberg/pull/4777/files/3cdf86f520f8f379b50eeed62872e24fb7516b76#r167880008 below, this would be something for a next iteration as well (since <Header/> is nested inside <Sidebar/>)

IreneStr committed:
Fix unittests

IreneStr committed:
Merge branch 'master' into add/api-add-plugin-sidebar-2

gziolo commented on edit-post/store/index.js:
```diff
@@ -8,7 +8,6 @@ import { registerReducer, withRehydratation, loadAndPersist } from '@wordpress/d
 import reducer from './reducer';
 import enhanceWithBrowserSize from './mobile';
-import applyMiddlewares from './middlewares';
 import { BREAK_MEDIUM } from './constants';
```
We don't need `.middlewares` anymore if we stop using it here.

@youknowriad can you confirm we can remove it completely?

youknowriad replied:
yes, it's a local dependency if the mobile middleware is gone we can safely remove it. but how is the mobile Middleware replaced right now? Sorry didn't review in depth?

jorgefilipecosta replied:
The need for mobile middleware was removed because of a logic was changed. As a side effect: 
> The sidebar is hidden when you change from desktop to mobile. However, it's not saved that it was opened before, which means it will not be reopened when switching back to desktop.

I think it is ok to remove that behavior as the reopen from mobile to desktop was not an initial requirement. And we can try to propose a solution for that if required after this PR is merged.

gziolo commented:
When I register my sidebar and activate it, then refresh the page I still see this error about missing plugin. I think it needs to be fixed, because it is expected that this sidebar shouldn't be there, so this error is only confusing for the user:

![screen shot 2018-02-19 at 16 42 26](https://user-images.githubusercontent.com/699132/36385988-9c393ad4-1594-11e8-8120-b1d1789e609e.png)

jasmussen commented:
While I really want this to get merged in, I would object to the merger [until these UI issues](https://github.com/WordPress/gutenberg/pull/4777#issuecomment-364047580) are fixed. Specifically:

- The plugin sidebar header is too tall compared to the post settings sidebar header
- The font is larger compared to the post settings sidebar header
- The X close button is offset compared to the post settings sidebar header

My primary concern is that we merge this in and work on addressing the UI glitches falls through the cracks, forgotten mostly because you'd only ever see it if you also installed a plugin that took advantage of this new API, which very few would do.

jorgefilipecosta commented:
The improvements I referred in the last review are addressed, thank you for applying the fixes.
The error @gziolo referred is caused by the plugin sidebar not being available when we reload. I referred this problem and said it may be future work because it probably happens when plugins are removed, but I was not aware of all the impact it may have in other cases e.g: plugin renaming sidebars.
So if we can add a verification to check if the sidebar is available when loading the preferences from local storage instead of showing the error, solving the reload issues @gziolo referred it would be nice.
Thank you for your contribution 👍

IreneStr committed:
Fix header height, font and close button

IreneStr commented:
@jasmussen My bad. I was under the impression that I had fixed those UI things when I did a CSS review and removed duplicate CSS. I've just pushed more CSS changes to address the issues you raised. 

@jorgefilipecosta @gziolo I'll address your feedback tomorrow.

jasmussen commented on edit-post/components/plugins-panel/style.scss:
```diff
@@ -12,6 +12,10 @@
     h3 {
       margin: 0;
+      font-weight: 400;
+      color: inherit;
```
Haven't had time to check, but should the font size also be addressed? The font size should be 13px, and you get that from the $default-font-size variable here https://github.com/WordPress/gutenberg/blob/master/edit-post/assets/stylesheets/_variables.scss#L7

IreneStr replied:
<img width="920" alt="schermafbeelding 2018-02-21 om 09 57 55" src="https://user-images.githubusercontent.com/17744553/36471160-bf7ecb76-16ed-11e8-8d1f-a265306b43a1.png">
@jasmussen The current font size is 13px. Or do I need to set it explicitly for some reason?

jasmussen replied:
Ah right, sorry, wasn't able to test. No that looks good. Though feel free to set the font weight to 600 so it's bold, like an active tab. Otherwise 👍 👍

gziolo committed:
Update index.js

gziolo committed:
Edit Post: Remove no longer used getPluginSidebar

gziolo committed:
Edit Post: Remove no longer used imports from PluginsPanel

gziolo committed:
Edit Post: Update Layout to handle missing plugin properly

gziolo commented:
I added a few commits to remove the sidebar state with error messages. Let's tackle it separately in a way that is less confusing for the users, but at the same time helpful for site owners so they could get proper feedback.

![screen shot 2018-02-19 at 16 42 26](https://user-images.githubusercontent.com/699132/36385988-9c393ad4-1594-11e8-8120-b1d1789e609e.png)

At this moment when there is no matching sidebar registered by Redux has active plugin stored, we stop rendering sidebar. Let's iterate on this in the follow-up PR.

gziolo committed:
Edit Post: Remove obsolete blank line

gziolo commented:
There is one known issue. Steps to reproduce:

* register the sidebar and activate it
* reload the page
* it gets closed as expected
* if then we register the sidebar again and activate it, it does not show up as expected

This is not likely that user will get into such flow, so I think we can fix it in a follow-up PR.

@IreneStr thanks for working on this one. It looks solid and we can proceed. We agreed that it's better to merge it in the current shape and address all the comments in the smaller targeted PRs. ✨
